code/go/0chain.net/smartcontract/dbs/event/eventdb.go

```diff
@@ -14,17 +20,79 @@ func NewEventDb(config dbs.DbAccess) (*EventDb, error) {
 		return nil, err
 	}
 	eventDb := &EventDb{
-		Store: db,
+		Store:          db,
+		eBufferChannel: make(chan eventCtx, 100),
+		eChannel:       make(chan eventCtx, 100),
 	}
 
 	if err := eventDb.AutoMigrate(); err != nil {
 		return nil, err
 	}
+	go eventDb.channelBufferIntermediate()
+	go eventDb.addEventWorker()
 	return eventDb, nil
 }
 
+type eventCtx struct {
+	ctx    context.Context
+	events []Event
+}
+
 type EventDb struct {
 	dbs.Store
+	eChannel       chan eventCtx
+	eBufferChannel chan eventCtx
+}
+
+// helps maitaining dynamic buffer for addEventWorker
+func (edb EventDb) channelBufferIntermediate() {
```

Comment:
A unit test would be nice. You could generate some semi-random events. Start channelBufferIntermediate. Send your test events to eBufferChannel. Monitor eChannel and confirm it receives events in correct order.

Reply:
UT brought out an edge case where the goroutine is not scheduled and the channel buffer is full
in this case, the order is still not guaranteed. 
Also, why cannot we use a queue service like kafka or rabbitmq to achieve this? Yes, it will increase our requirement but if that is what we want to achieve this is the only solution.
If this was a Kubernetes-based solution. I would have recommended using new-relic to log all the events.
Please read the following golang issue where someone tried to do this
https://github.com/golang/go/issues/20352

Reply:
I don't follow, maybe you could commit a unit test that exposes the problem so we can see what you mean.

Reply:
Semaphores, mutexes, monitors, etc, in general knows nothing about data structure they serialize access to. channels is something different, it's more like data structure and semaphore together.

As I got your problem right, you can serialize access to a slice with mutext and implement FIFO logic with this slice easily, you can do queue with two slices (copy on empty) it will be more performant in some cases.
What operations you expect to prevail, reads or writes?

Reply:
@dabasov no the issue is if the goroutine (that appends to the slice) does not get scheduled and the buffer for channel is full
Then the events are not serialized
cc: @Sriep

Reply:
```go
	eventDb := &EventDb{
		Store:          db,
		eBufferChannel: make(chan eventCtx, 100),
		eChannel:       make(chan eventCtx, 100),
	}
```
Why is eBufferChannel buffered.
```go
func (edb *EventDb) AddEvents(ctx context.Context, events []Event) {
	edb.eBufferChannel <- eventCtx{
		ctx:    ctx,
		events: events,
	}
}
```
Don't we want this to be synchronous? We are only adding the events to `buf` so this should not hold up the chain. 

Making it `eBufferChannel` buffered raises the issue if the buffer gets full. Also, we should probably change the name as we don't want it buffered.

For the moment I think we can skip the issue with running out of memory with an oversized `buf` object if the event database gets stuck. We can relook if it becomes an issue.